pkg/agent/nodeportlocal/nplagent_test.go

```diff
@@ -0,0 +1,326 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodeportlocal
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/k8s"
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+	"github.com/vmware-tanzu/antrea/pkg/signals"
+
+	npltest "github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/rules/testing"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sfake "k8s.io/client-go/kubernetes/fake"
+)
+
+func NewPortTable(c *gomock.Controller) *portcache.PortTable {
+	ptable := portcache.PortTable{StartPort: 40000, EndPort: 45000}
+	ptable.Table = make(map[int]portcache.NodePortData)
+
+	mockTable := npltest.NewMockPodPortRules(c)
+	mockTable.EXPECT().AddRule(gomock.Any(), gomock.Any()).AnyTimes()
+	mockTable.EXPECT().DeleteRule(gomock.Any(), gomock.Any()).AnyTimes()
+
+	ptable.PodPortRules = mockTable
+	return &ptable
+}
+
+const (
+	defaultPodName  = "test-pod"
+	defaultNS       = "default"
+	defaultNodeName = "test-node"
+	defaultHostIP   = "10.10.10.10"
+	defaultPodIP    = "192.168.32.10"
+	defaultPort     = 80
+)
+
+var kubeClient *k8sfake.Clientset
+var portTable *portcache.PortTable
+
+func getTestPod() corev1.Pod {
+	testPod := corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      defaultPodName,
+			Namespace: defaultNS,
+		},
+		Spec: corev1.PodSpec{
+			NodeName: defaultNodeName,
+			Containers: []corev1.Container{
+				{
+					Ports: []corev1.ContainerPort{
+						{
+							ContainerPort: int32(defaultPort),
+						},
+					},
+				},
+			},
+		},
+		Status: corev1.PodStatus{
+			HostIP: defaultHostIP,
+			PodIP:  defaultPodIP,
+		},
+	}
+	return testPod
+}
+
+func TestMain(t *testing.T) {
+	os.Setenv("NODE_NAME", defaultNodeName)
+	kubeClient = k8sfake.NewSimpleClientset()
+	mockCtrl := gomock.NewController(t)
+	portTable = NewPortTable(mockCtrl)
+
+	c, _ := InitController(kubeClient, portTable, defaultNodeName)
+	stopCh := signals.RegisterSignalHandlers()
+	go c.Worker()
+	go c.Ctrl.Run(stopCh)
+}
+
+// Add a new Pod with fake k8s client and verify that npl annotation gets updated
+// and an entry gets added in the local port cache
+func TestPodAdd(t *testing.T) {
+	var ann map[string]string
+	var data string
+	var found bool
+
+	a := assert.New(t)
+
+	testPod := getTestPod()
+	p, err := kubeClient.CoreV1().Pods(defaultNS).Create(context.TODO(), &testPod, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Pod creation failed with err: %v", err)
+	}
```

Comment:
for less verbose tests, you should import `"github.com/stretchr/testify/require"` and use:

```go
require.Nil(err, "Pod creation failed")
```

same for other error checks below